src/plugins/preauth/pkinit/pkinit_srv.c

```diff
@@ -1489,3 +1567,316 @@ kdcpreauth_pkinit_initvt(krb5_context context, int maj_ver, int min_ver,
+            /* Assert authentication indicators from certificate authorization. */
+            for (ai = ais; *ai != NULL; ai++) {
+                ret = cb->add_auth_indicator(context, rock, *ai);
+                if (ret)
+                    goto cleanup;
+            }
+
+            if (hd->vt.free_ind != NULL)
+                hd->vt.free_ind(context, hd->moddata, ais);
+            else
+                free_ai_list(ais);
+            ais = NULL;
+        }
+    }
+
+    ret = accepted ? 0 : KRB5KDC_ERR_CLIENT_NAME_MISMATCH;
+
+cleanup:
+    crypto_free_cert_data(context, cert);
+    return ret;
+}
+
+static krb5_error_code
+pkinitsrv_san_authorize(krb5_context context, krb5_certauth_moddata moddata,
+                        const uint8_t *cert, size_t cert_len,
+                        krb5_const_principal princ, const void *opts,
+                        const krb5_db_entry *db_entry, char ***authinds_out)
+{
+    krb5_error_code ret;
+    int valid_san;
+    const struct certauth_req_opts *req_opts = opts;
+
+    if (authinds_out != NULL)
+        *authinds_out = NULL;
+
+    /* Verify the client SAN. */
+    ret = verify_client_san(context, req_opts->plgctx, req_opts->reqctx,
+                            req_opts->cb, req_opts->rock, princ, &valid_san);
+    if (ret == ENOENT) {
+        ret = KRB5_PLUGIN_NO_HANDLE;
+        goto out;
+    }
+
+    if (ret)
+        goto out;
+
+    if (!valid_san) {
+        pkiDebug("%s: did not find an acceptable SAN in user "
+                 "certificate\n", __FUNCTION__);
+        ret = KRB5KDC_ERR_CLIENT_NAME_MISMATCH;
+    }
+
+out:
+    return ret;
+}
+static krb5_error_code
+pkinitsrv_eku_authorize(krb5_context context, krb5_certauth_moddata moddata,
+                        const uint8_t *cert, size_t cert_len,
+                        krb5_const_principal princ, const void *opts,
+                        const krb5_db_entry *db_entry, char ***authinds_out)
+{
+    krb5_error_code ret;
+    int valid_eku;
+    const struct certauth_req_opts *req_opts = opts;
+
+    if (authinds_out != NULL)
+        *authinds_out = NULL;
+
+    /* Verify the client EKU. */
+    ret = verify_client_eku(context, req_opts->plgctx, req_opts->reqctx,
+                            &valid_eku);
+    if (ret)
+        goto out;
+
+    if (!valid_eku) {
+        pkiDebug("%s: did not find an acceptable EKU in user "
+                 "certificate\n", __FUNCTION__);
+        ret = KRB5KDC_ERR_INCONSISTENT_KEY_PURPOSE;
+    }
+
+out:
+    return ret;
+}
+
+static krb5_error_code
+certauth_pkinitsrv_san_initvt(krb5_context context, int maj_ver, int min_ver,
+                              krb5_plugin_vtable vtable)
+{
+    krb5_certauth_vtable vt;
+
+    if (maj_ver != 1)
+        return KRB5_PLUGIN_VER_NOTSUPP;
+    vt = (krb5_certauth_vtable)vtable;
+    vt->name = "pkinitsrv";
+    vt->authorize = pkinitsrv_san_authorize;
+    vt->init = NULL;
+    vt->fini = NULL;
+    vt->free_ind = NULL;
```

Comment:
It's not our practice to explicitly set methods to NULL in initvt functions; not having to do so is a nice benefit of using a zero-filled vtable provided by the caller.  (Also applies to the EKU module, the test1 module, and the dbmatch module.)